File: kotlin-orders/kotlin-orders-dto/src/main/java/com/mvv/bank/orders/repository/jpa/conversion/FxRateMapper.kt
package com.mvv.bank.orders.repository.jpa.conversion

import com.mvv.bank.orders.conversion.DomainPrimitiveMappers
import com.mvv.bank.orders.conversion.MAP_STRUCT_COMPONENT_MODEL
import com.mvv.bank.orders.domain.CurrencyPair
import com.mvv.bank.orders.domain.MarketSymbol
import com.mvv.bank.orders.domain.FxRate as DomainFxRate
import com.mvv.bank.orders.repository.jpa.entities.FxRate as JpaFxRate
import org.mapstruct.Mapper
import org.mapstruct.Mapping


@Mapper(componentModel = MAP_STRUCT_COMPONENT_MODEL, config = DomainPrimitiveMappers::class)
interface FxRateMapper {

    fun fromDto(source: JpaFxRate?): DomainFxRate? =
        if (source == null) null else DomainFxRate(
            MarketSymbol(source.market), source.timestamp, source.marketDate, source.marketTime,
            CurrencyPair.of(source.cur1, source.cur2), bid = source.bid, ask = source.ask)


    @Mapping(source = "currencyPair.base", target = "cur1")
    @Mapping(source = "currencyPair.counter", target = "cur2")
    fun toDto(source: DomainFxRate?): JpaFxRate?
}

File: scala-orders/src/main/scala/com/mvv/bank/orders/rest/conversion/FxRateMapper.scala
package com.mvv.bank.orders.rest.conversion

import org.mapstruct.{BeanMapping, Mapper, Mapping}
import com.mvv.bank.orders.domain.FxRate as DomainFxRate
import com.mvv.bank.orders.domain.{CurrencyPair as DomainCurrencyPair, MarketSymbol as DomainMarketSymbol}
import com.mvv.bank.orders.rest.entities.FxRate as DtoFxRate


// TODO: Unmapped target property: "currencyPair". Occured at 'FxRate fxRateToDomain(FxRate source)' in 'FxRateMapper'
// TODO: Unmapped target properties: "cur1, cur2". Occured at 'FxRate fxRateToDto(FxRate source)' in 'FxRateMapper'


//@Mapper
trait FxRateMapper :
  //@Mapping(target = "currencyPair", expression = "CurrencyPair.of(source.cur1(), source.cur2()) ")
  //def fxRateToDomain(source: DtoFxRate): DomainFxRate
  def fxRateToDomain(source: DtoFxRate): DomainFxRate =
    DomainFxRate(
      market = DomainMarketSymbol(source.market),
      timestamp = source.timestamp,
      marketDate = source.marketDate,
      marketTime = source.marketTime,
      currencyPair = DomainCurrencyPair(source.cur1, source.cur2),
      bid = source.bid,
      ask = source.ask,
    )
  //def fxRateCurrencyPair(source: DtoFxRate): DomainCurrencyPair =
  //  DomainCurrencyPair(source.cur1, source.cur2)
  def fxRateToDomain(source: Option[DtoFxRate]): Option[DomainFxRate] = source.map(fxRateToDomain)

  @Mapping(target = "cur1", source = "currencyPair.base")
  @Mapping(target = "cur2", source = "currencyPair.counter")
  //@Mapping(target = "mid", ignore = true)
  @BeanMapping(ignoreUnmappedSourceProperties = Array("mid"))
  def fxRateToDto(source: DomainFxRate): DtoFxRate
  def fxRateToDto(source: Option[DomainFxRate]): Option[DtoFxRate] =
    source.map(fxRateToDto)
